### DISCUSSION TITLE
Compiling app with pyinstaller on M1 Mac and universal2 architecture

### DISCUSSION THREAD

**Discussion by willi3by** [pyinstaller/pyinstaller#7323] [Help] [answered]
Hi,

I am brand new to pyinstaller and am trying to compile an app that uses PyQt5 and PyMuPDF to send to a colleague. He has a set of PDFs that need to be redacted before I can see them so I need an easy to use app that works on his computer so he can redact them. The app that I have works as expected and compiles as expected on my M1 Mac, but when I try to compile for intel and M1 (universal2), I get the following error message:

```
PyInstaller.utils.osx.IncompatibleBinaryArchError: /opt/homebrew/Caskroom/miniforge/base/envs/pyqt_py3.9/lib/python3.9/lib-dynload/math.cpython-39-darwin.so is not a fat binary!
```

Every time I run the command, I get the same error but it refers to a different library, e.g., `fitz.cpython-39-darwin.so` as opposed to `math.cpython-39-darwin.so`. I understand what the error message is saying but I am unsure if there is a way around this problem. The person I am sending the app to has the same MacOS (12.6) but has an Intel chip instead of an M1, so I am hoping there is still some way to compile for his computer. 

I am using:
Pyinstaller: 5.6.2
Python: 3.9.10
PyQt: 5.15.4

Any guidance is very much appreciated. Please let me know if there is any other info that would be useful. 

Thanks

**Comment by rokm** [comment, marked as answer]
>I understand what the error message is saying but I am unsure if there is a way around this problem.

There is not. You cannot create `universal2` builds using a single-arch python environment. If you want to create `universal2` build, your python *and* all non-pure-python packages (like PyQt5) need to be `universal2`. That rules out homebrew (which is always single-arch), and may also rule out python.org python (which is universal2) if PyQt5 PyPI wheels are not universal2 (last I checked, they were x86_64-only).

So I suppose the only option left in this scenario is to set up x86_64 python environment on your M1 mac (which will run under Rosetta emulation), and create a single-arch x86_64 build. Which can then run natively on your colleague's Intel Mac, and under Rosetta on your M1 Mac.

**Reply by willi3by** [reply]
Thanks for the quick answer! That helped a lot, not only with this project, but a few others as well. I had not thought of Rosetta emulation and had no idea I could run GUIs made with PyQt from a Rosetta terminal. The app is now working as expected with both arm64 and x86_64 builds.
